pages/docs/couple-your-code/couple-your-code-adapter-software-engineering.md

```diff
@@ -2,17 +2,119 @@
 
-The direct modification approach is what the [step-by-step guide](couple-your-code-preparing-your-solver) uses. It consists of directly modifying the solver code lines to couple with preCICE. This, however, is not an ideal approach since it requires changing of the solver source code, and you should try to avoid this to wherever possible to maintain a sustainable software development practice.
+{% tip %}
+In any case, fully integrating the adapter into the codebase and the continuous integration system of the solver (with any of the methods described here) would make maintenance easier, as most complications typically come from changes in the solver code, not from preCICE (in which case, you can refer to a [porting guide](./couple-your-code-porting-overview.html)).
+{% endtip %}
 
-To minimize the lines of solver source code changes, you can create a separate class for the adapter. This adapter will be responsible for calling all the preCICE APIs, and from the solver source code you would only call the corresponding adapter class methods. As stated above, this approach is used e.g. for the SU2-adapter and explained in detail in [Alexander Rusch's thesis](https://www5.in.tum.de/pub/Rusch2016_BA.pdf).
+### Adapter class
 
-If the solver you are using provides a callback functionality you can separate the adapter even better from the code by calling the preCICE API from the callback. As stated above this is realized in the [OpenFOAM adapter](https://github.com/precice/openfoam-adapter) and is explained in detail in [Gerasimos Chourdakis' thesis](https://www5.in.tum.de/pub/Chourdakis2017_Thesis.pdf).
+Encapsulating the adapter into a class separates the preCICE calls from the rest of the solver code.
+This is very similar to directly modifying the solver code (with some modifications still needed),
+but the solver calls a few high-level adapter functions, instead of directly calling the preCICE API.
+Such encapsulation avoids duplicating additional checks and details and makes maintenance easier.
+
+This method is particularly useful for adapters used in frameworks that the user uses to create concrete simulations.
+Examples here include the [FEniCS adapter](./adapter-fenics.html) and the [deal.II adapter](https://github.com/precice/dealii-adapter).
+
+Even when directly modifying the code, it makes sense to hide as many changes as possible behind a class API.
+Examples for this include the [CalculiX adapter](./adapter-calculix-overview.html) and the [SU2 adapter for v6 and earlier](./adapter-su2-overview.html) (as described in [Alexander Rusch's thesis](https://www5.in.tum.de/pub/Rusch2016_BA.pdf)).
+
+Adapter classes can still be provided as optional modules of the main codebase. Examples include the [deal.II adapter](./adapter-dealii-overview.html), the [G+smo adapter](./adapter-gismo.html), the [DUNE adapter](./adapter-dune.html), and the [DuMux adapter](./adapter-dumux-get.html)
```

Comment:
Would it be easier to understand if we bring the packaging aspect earlier in this section? "Separate class", but then one still has different option on how to maintain and package, "integrated" or "independent"?